compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp

```diff
@@ -459,13 +457,184 @@ struct DistributeBroadcast final : OpDistributionPattern<vector::BroadcastOp> {
+
+    if (!locallyReduced) {
+      return rewriter.notifyMatchFailure(
+          localReduction, "unimplemented: scalar result distribution");
+    }
+
+    // Flatten the locally reduced value.
+    VectorType shaped = locallyReduced.getType();
+    int64_t numElements = shaped.getNumElements();
+    SmallVector<int64_t> flatShape(1, numElements);
+    Type elemTy = shaped.getElementType();
+    VectorType flatVecType = VectorType::get(flatShape, elemTy);
+    VectorValue flat =
+        rewriter.create<vector::ShapeCastOp>(loc, flatVecType, locallyReduced);
+
+    FailureOr<VectorValue> globalReduced = doThreadGlobalReduction(
+        rewriter, srcLayout, flat, multiReduceOp.getKind(), reducedDims);
+    if (failed(globalReduced)) {
+      return failure();
+    }
+
+    VectorValue unflattened = rewriter.create<vector::ShapeCastOp>(
+        loc, shaped, globalReduced.value());
+    replaceOpWithDistributedValues(rewriter, multiReduceOp, unflattened);
+
+    return failure();
+  }
+
+  FailureOr<VectorValue>
+  doThreadGlobalReduction(RewriterBase &rewriter, NestedLayoutAttr layout,
+                          VectorValue flat, vector::CombiningKind kind,
+                          ArrayRef<bool> reductionMask) const {
+    VectorType flatVecType = flat.getType();
+    Type elemTy = flatVecType.getElementType();
+    int64_t numElements = flatVecType.getNumElements();
+    Location loc = flat.getLoc();
+
+    unsigned elemBitwidth = elemTy.getIntOrFloatBitWidth();
+    if (elemBitwidth != maxBitsPerShuffle) {
+      return rewriter.notifyMatchFailure(
+          flat.getDefiningOp(), llvm::formatv("unimplemented: packed shuffle",
+                                              elemBitwidth, maxBitsPerShuffle));
+    }
+
+    auto constOp = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(flatVecType));
+    auto res = llvm::cast<VectorValue>(constOp.getResult());
+
+    for (unsigned i = 0; i < numElements; ++i) {
+      Value extracted = rewriter.create<vector::ExtractOp>(loc, flat, i);
+
+      // Reduce across all reduction dimensions 1-by-1.
+      for (unsigned i = 0; i < reductionMask.size(); ++i) {
+        if (reductionMask[i]) {
+          extracted = doPackedThreadReductionOnDim(rewriter, layout, extracted,
+                                                   kind, i);
+        }
+      }
+
+      res = rewriter.create<vector::InsertOp>(loc, extracted, res, i);
+    }
+
+    return res;
+  }
+
+  Value doPackedThreadReductionOnDim(RewriterBase &rewriter,
+                                     NestedLayoutAttr layout, Value val,
+                                     vector::CombiningKind kind,
+                                     int64_t dim) const {
+    Location loc = val.getLoc();
+    int64_t offset = getShuffleOffset(layout, dim);
+    int64_t width = getShuffleWidth(layout, dim);
+
+    // TODO: Query from entrypoint.
+    int64_t subgroupSize = 64;
```

Comment:
Can you add this in this PR? I have a feeling this may byte us if we ever try to enable vector distribution on a new target.

Reply:
+1, I've done the same before and it ends up being more hassle than it's worth, so doing the extra plumbing up front is much appreciated. FWIW we'll need to change this pretty soon for gfx1100 where we are using subgroup size 32 right now.

Reply:
I plumbed the subgroup size to ContractionVectorLayoutOptions. Not sure how to plumb it further.

Reply:
Nice, all you should need now is just retrieving the subgroup size from the enclosing function. You can get the subgroup size with this helper: https://github.com/iree-org/iree/blob/afd7cab58be44c035eb80aaad39652b4f5d5a667/compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h#L51

Reply:
Done for vector distribution pass. For transform dialect stuff though i hardcoded the subgroup size for now.